Short-Answer/Algorithms_Answers.md

```diff
@@ -2,14 +2,61 @@
 
 ## Exercise I
 
-a)
+a). First line is setting `a` to zero. This is a constant time operation. Setting memory to some new value 0(1)
 
+    Second Line `while  (a < n * n * n):` This while loop will continue until the condiditon is true. In this case its `a < n * n * n`. Comparing the variable a to n^3. As long as a is less than n * n * n the loop will continue giving this an operation od 0(1).
 
-b)
+    third Line: is setting a to a new value and increasing by one. This operation is 0(1) because it pretty simular to the first line
 
+    0(1) + 0(2) + 0(1) = 0(n2)
 
-c)
+
+
+b) First Line: We are setting a counter called sum and equaling that to 0. We are hard coding the value and this is hapening in constant time with a operation of 0(1)
+    Second line: We are traversing through the list on `n`. We are going through every single element until we get to the end. This happens more thanonces and its changing the value of n, giving this a operation of 0(n)
+
+    third Line: once again we are setting a variable to a new value. This always will be a constant time operation of 0(1)
+
+    forth line: here is another loop. This loop is compaing the value of j to the lenght of `n`. If this remains true the while loop will keep excuting until it is false. `while j < n:` this is an 0(Log n) operation because of the loop and compaing the values.
+
+    fifth line is taking j and multiplying the results. This only happens if `j` is less than `n`if it is j will multiple by two(j = j * 2 ). So the data is changing as long as this condiditon remains true. 
+    0(2) operation 
+
+    six line: this is an 0(1) operation becasue we are taking the sum variable and adding one to it.
+
+    0(1) + 0(n) * 0(1) + 0(Log n) * 0(1) = 0(log n2)
+
+c) first line if compaing the n value and 0. 0(1) operation
+
+    Second line: return 0 if bunnies == 0 0(1) operation
+
```

Comment:
Check out some videos on time complexity. We're less focused on the time it takes to assign a variable, and more focused on the time it takes to progress through a loop.
```py
my_sum = 0
for i in range(n):
  my_sum += i
```

This loop for example. Creating and updating the values for `my_sum` are all `O(1)` operations. But, what is the time complexity of the for loop itself?

This about those types of questions - not just the time to assign values, but the time it takes to traverse through a full loop.

Reply:
Im not sure if your read my answers but i am talking about the loops and the time complexity in each answer along with the variables. If your guys only wanted us to focus on JUST the LOOPS then that should of been stated in the readme or directions.

Reply:
Yes, I read all of your answers thoroughly. Being aware and able to recognize `O(1)` time complexities is good! But when you're looking at the time complexity of a larger piece of code, `O(1)` doesn't come into play.

There are a few things I noticed from these answers you gave, and I chose to only point out the loops. Because it seems like you want more critique on your full answer, I'll elaborate more. I won't give you the answers, but I'll identify the points where I think you may be struggling with time complexity.

## Example 1

```py
my_sum = 0
for i in range(n*n):
    my_sum += 1
```
Take this loop for example. Like you've answered in the sprint, lines 1 and 3 have a time complexity of `O(1)`. You could even argue that line 2 has a time complexity of `O(1)` because `range(n * n)` is a single operation, and assigning and reassigning `i` is a single operation too.

If this is all you use to find the time complexity however, you will get something like
```
O(1) + O(1) + O(1) = O(3)
```
And that's not correct.

To get the time complexity of the full loop, you have to look at _how many times you're looping overall._ In this case, we're looping `n * n` times, because that's our range to loop through. Because this is a simple for loop, we know we're going to go through every integer in our range, and we won't skip over any numbers.

We can also use examples to help us find time complexity:
* With an input of 3, we would loop 9 times.
* With an input of 4, we would loop 16 times.
* With an input of 5, we would loop 25 times.

### Final Time complexity
Looking at this, we can determine that our time complexity is `O(n*n)`. Every time our input grows, if the input is `n` then we loop `n*n` times. The time it takes to assign and reassign the variables doesn't change this final time complexity.

## Example 2

Here's one more example.
```py
for i in range(n):
  for j in range(5):
    print(i + j)
```
To find the time complexity here, we still want to focus on how many times we need to loop as our input changes. Line three is an `O(1)` operation, but knowing that alone won't help us find the solution. To break this down:
* We have one loop that loops `n` times
* Inside that loop, we have a nested loop that always loops 5 times
* We print (an `O(1)` operation stated above)

The quick and dirty way to find the time complexity of this is to first find the time complexity of each loop. Because they're nested, we can **multiply** the ranges together and find our final time complexity.
* the first loop is `O(n)` because it increments 1 by 1 until it reaches `n`.
* the first loop is `O(5)` because it increments 1 by 1 until it reaches 5. Changing `n` will never change this loop because it'll always go up to 5.

Now that we have our time complexities of each loop, we can **multiply the complexities together because they are nested loop:**
```
O(n) * O(5) = O(5n) = O(n)
Remember, we take out  ↑↑↑ any simple numbers when we calculate time complexity
```
We could also test by passing in actual numbers for `n` like we did above:
* With an input of 3, we would print 15 times.
* With an input of 4, we would print 20 times.
* With an input of 5, we would print 25 times.
* With an input of 6, we would print 30 times.
As our input increases, our loops increase by 5 every time.

### Final Time Complexity
This second example would have a time complexity of `O(n)`. Technically, it has a time complexity of `O(5n)`, but remember - we take off any full numbers that are multiplied or added in our time complexity.

## Conclusion
Looking at the questions in this challenge, ask yourself:
* What is the time complexity of each loop?
* Are the loops nested?
    * If they are, you can multiply the time complexities.
    * If they aren't, you just add the time complexities.
    * `O(5n)` versus `O(5 + n)`
* If you pass in an input, see how many times you loop and see if you can find a pattern.

Again, the issue is not that we want you to only look at the loops, but rather _that's just how time complexity works._ This is why I recommend looking up videos to explain time complexity. You explained a lot in these answers, and some of what you said is correct, and good information. But the final answers you have are not correct, and when it comes down to more complex loops, your explanations are also incorrect. With final time complexities overall, analyzing each line 1 by 1 doesn't always help.

**And an extra aside - If someone is doing a code review on your work, it is disrespectful to tell _them_ what _they should have done_.**
Asking questions for a deeper understanding is good.
Clarifying information given to you is also good. 
Make sure you Understand the challenge before you critique it.

Most people would not take the extra time for this longer explanation.